### PR TITLE
feat: add livery conversion

### PR DESCRIPTION
**Pull request by Benjozork**
## Summary of Changes

* Add livery conversion utility (LCU)
* Add incompatible livery warning
* Add setting to ignore livery warning

Discord username (if different from GitHub): someperson#4953

### CODE REVIEW

File: src/renderer/components/AircraftSection/LiveryConversion/index.tsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            entries
+                .filter((entry) => entry.state !== LiveryState.ERROR_DURING_CONVERSION)
+                .forEach((entry) => store.dispatch<SetLiveryStateAction>({
+                    type: actionTypes.SET_LIVERY_STATE,
+                    payload: {
+                        livery: entry.livery,
+                        state: LiveryState.TO_BE_CONVERTED,
+                    },
+                }));
+        }
```
Comment (nistei) [inline src/renderer/components/AircraftSection/LiveryConversion/index.tsx:172]: Try and prevent code duplication.

File: src/renderer/components/AircraftSection/styles.tsx [hunk 2]
```diff
@@ -120,9 +120,14 @@ export const Content = styled.div`
     padding: 6px 24px 24px 24px;
 `;
 
-export const TopContainer = styled.div`
+export const ModalContainer = styled.div`
```
Comment (nistei) [inline src/renderer/components/AircraftSection/styles.tsx:123]: Again, not a modal dialog

File: src/renderer/utils/LiveryConversion.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+export type LiveryDefinition = {
+    packageName: string,
+    simObjectName: string
+    title: string,
+}
+
+export class LiveryConversion {
+
+    static async getIncompatibleLiveries(): Promise<LiveryDefinition[]> {
+        return (promisify(fs.readdir)(Directories.community()))
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:17]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    private static async convertPackageManifest(packageFolder: string): Promise<void> {
+        const manifestPath = path.join(packageFolder, 'manifest.json');
+        const manifestContents = (await promisify(fs.readFile)(manifestPath)).toString();
+
+        const manifest = JSON.parse(manifestContents);
+        manifest['title'] = `${manifest['title']} (A32NX Converted)`;
+
+        await promisify(fs.writeFile)(manifestPath, JSON.stringify(manifest));
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:166]: `fs-extra` offers a promise function `writeJSON()`.

File: src/renderer/utils/LiveryConversion.ts [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    private static async convertPackageLayout(packageFolder: string, oldSimObjectName: string, newSimObjectName: string): Promise<void> {
+        const layoutPath = path.join(packageFolder, 'layout.json');
+        const layoutContents = (await promisify(fs.readFile)(layoutPath)).toString();
+
+        await promisify(fs.writeFile)(layoutPath, layoutContents.replaceAll(oldSimObjectName, newSimObjectName));
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:173]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    private static async convertAircraftCfg(packageFolder: string, simObjectName: string): Promise<{ textureFolderName: string, modelFolderName: string }> {
+        const parser = new ConfigIniParser();
+
+        const aircraftCfgPath = path.join(packageFolder, 'SimObjects', 'AirPlanes', simObjectName, 'aircraft.cfg');
+        const aircraftCfgContents = (await promisify(fs.readFile)(aircraftCfgPath)).toString();
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:180]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+                    return Promise.reject('Livery does not have a \'model\' folder');
+                }
+            } else {
+                return Promise.reject('No \'model\' option found in [FLTSIM.0] section');
+            }
+        } else {
+            return Promise.reject('No [FLTSIM.0] section found. Only [FLTSIM.0] is currently supported');
+        }
+
+        await promisify(fs.writeFile)(aircraftCfgPath, aircraftCfg.stringify());
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:251]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+        return { textureFolderName, modelFolderName };
+    }
+
+    private static async convertTextureCfg(packageFolder: string, simObjectName: string, textureFolderName: string): Promise<void> {
+        const parser = new ConfigIniParser();
+
+        const textureCfgPath = path.join(packageFolder, 'SimObjects', 'AirPlanes', simObjectName, `TEXTURE.${textureFolderName}`, 'texture.cfg');
+        const textureCfgContents = (await promisify(fs.readFile)(textureCfgPath)).toString();
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:260]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+                continue;
+            }
+
+            const value = textureCfg.get('fltsim', option) as string;
+            if (value.includes('Asobo_A320_NEO')) {
+                textureCfg.set('fltsim', option, value.replace('Asobo_A320_NEO', 'FlyByWire_A320_NEO'));
+            }
+        }
+
+        await promisify(fs.writeFile)(textureCfgPath, textureCfg.stringify());
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:279]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    private static async convertModelCfg(packageFolder: string, simObjectName: string, modelFolderName: string): Promise<void> {
+        const parser = new ConfigIniParser();
+
+        const modelCfgPath = path.join(packageFolder, 'SimObjects', 'AirPlanes', simObjectName, `MODEL.${modelFolderName}`, 'model.cfg');
+        const modelCfgContents = (await promisify(fs.readFile)(modelCfgPath)).toString();
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:286]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+
+                modelCfg.set(
+                    'models',
+                    'interior',
+                    originalValue.replaceAll('Asobo_A320_NEO', 'FlyByWire_A320_NEO')
+                );
+            }
+        }
+
+        await promisify(fs.writeFile)(modelCfgPath, modelCfg.stringify());
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:315]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        console.log(`[LCU/Conversion] Converted model.cfg in '${newSimObjectName}'...`);
+
+        console.log(`[LCU/Conversion] Done converting '${newSimObjectName}'.`);
+
+        return true;
+    }
+
+    private static async convertPackageManifest(packageFolder: string): Promise<void> {
+        const manifestPath = path.join(packageFolder, 'manifest.json');
+        const manifestContents = (await promisify(fs.readFile)(manifestPath)).toString();
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:161]: `fs-extra` is already in the packages and allows for using promises.

File: src/renderer/utils/LiveryConversion.ts [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+        const manifest = JSON.parse(manifestContents);
+        manifest['title'] = `${manifest['title']} (A32NX Converted)`;
+
+        await promisify(fs.writeFile)(manifestPath, JSON.stringify(manifest));
+    }
+
+    private static async convertPackageLayout(packageFolder: string, oldSimObjectName: string, newSimObjectName: string): Promise<void> {
+        const layoutPath = path.join(packageFolder, 'layout.json');
+        const layoutContents = (await promisify(fs.readFile)(layoutPath)).toString();
```
Comment (nistei) [inline src/renderer/utils/LiveryConversion.ts:171]: `fs-extra` is already in the packages and allows for using promises.

**Review by nistei — CHANGES_REQUESTED** [review]

File: src/renderer/components/AircraftSection/LiveryConversion/index.tsx [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+import { LiveryConversion, LiveryDefinition } from "renderer/utils/LiveryConversion";
+import { AlertTriangle, Check } from "tabler-icons-react";
+import { Progress } from "antd";
+import { LiveryState, LiveryStateEntry } from "renderer/redux/reducers/liveries.reducer";
+import { LiveryAction, SetLiveryStateAction } from "renderer/redux/types";
+import Store from "electron-store";
+
+const settings = new Store();
+
+export const LiveryConversionModal: React.FC = () => {
```
[the remark this answers is not in the document]
Comment (FoxtrotSierra6829) [reply src/renderer/components/AircraftSection/LiveryConversion/index.tsx:14]: yeah was just about to write the same XD

File: src/renderer/components/AircraftSection/LiveryConversion/index.tsx [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+                        <a className="text-xl text-white font-semibold mr-3.5" onClick={handleDontAskAgain}>
+                            Don't Ask Again
+                        </a>
+                    }
+                    <button
+                        disabled={buttonDisabled()}
+                        className={`w-40 bg-red-500 transition-colors duration-200 hover:bg-red-700 px-6 py-2 text-2xl rounded-lg text-white ` + (buttonDisabled() ? 'pointer-events-none opacity-60' : '')}
+                        onClick={showList ? handleConfirm : handleConvert}
+                    >
+                        {showList ? 'Confirm' : 'Convert...'}
```
Comment (nistei) [inline src/renderer/components/AircraftSection/LiveryConversion/index.tsx:105]: The `Convert...` throws me off and looks to me like the text is cut off on the button.

```suggestion
                        {showList ? 'Confirm' : 'Convert'}
```

**Review by nistei — CHANGES_REQUESTED** [review]

**Comment by tracernz** [comment]
The converter needs to set `ui_manufacturer` and `ui_type` to match https://github.com/flybywiresim/a32nx/blob/master/flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/aircraft.cfg#L78-L79, otherwise there will be a mix-up with the Asobo A320neo.

**Comment by Benjozork** [comment]
> 
> 
> ~The converter needs to set `ui_manufacturer` and `ui_type` to match https://github.com/flybywiresim/a32nx/blob/master/flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/aircraft.cfg#L78-L79, otherwise there will be a mix-up with the Asobo A320neo.~
> 
> This should help: [flybywiresim/a32nx#4303](https://github.com/flybywiresim/a32nx/pull/4303). As long as we ensure `ui_manufacturer` is not set to something other than `TT:AIRCRAFT.UI_MANUFACTURER` or `FlyByWire Simulations` and `ui_type` `TT:AIRCRAFT.UI_MODEL` or `A320neo (LEAP)`.

Do you think we could ensure all converted liveries do ?

**Review by nistei — APPROVED** [review]
LGTM

File: src/renderer/utils/LiveryConversion.ts [hunk 16]
```diff
@@ -281,6 +277,9 @@ export class LiveryConversion {
         const parser = new ConfigIniParser();
 
         const modelCfgPath = path.join(packageFolder, 'SimObjects', 'AirPlanes', simObjectName, `MODEL.${modelFolderName}`, 'model.cfg');
+        if (!existsSync(modelCfgPath)) {
+            return null;
```
Comment (Benjozork) [inline src/renderer/utils/LiveryConversion.ts:281]: No need to return anything, it's a void function

Comment (Benjozork) [reply src/renderer/utils/LiveryConversion.ts:281]: @FoxtrotSierra6829

Comment (FoxtrotSierra6829) [reply src/renderer/utils/LiveryConversion.ts]: resolved @Benjozork

**Pull request merged**